File: Ryan Bullough/week04/Wednesday/MTA.rb
$lines = {
  N: ["Times Square", "34th", "28th", "23rd", "Union Square", "8th"],
  L: ["8th", "6th", "Union Square", "3rd", "1st"],
  "6": ["Grand Central", "33rd", "23rd", "Union Square", "Astor Place"],
}

def plan_trip
  print "Enter starting line \nN/L/6: "
  start_line_input = gets.chomp.upcase.to_sym
  starting_line = $lines[start_line_input]
  union_start = starting_line.index "Union Square"
  
  print "Enter starting station \n#{ starting_line.join(', ') }: "
  starting_station = gets.chomp.split.map(&:capitalize).join(' ')
  start_index = starting_line.index starting_station
  
  print "Enter destination line \nN/L/6: "
  destination_line_input = gets.chomp.upcase.to_sym
  destination_line = $lines[destination_line_input]
  union_end = destination_line.index "Union Square"
  
  print "Enter destination station \n#{ destination_line.join(', ') }: "
  destination_station = gets.chomp.split.map(&:capitalize).join(' ')
  destination_index = destination_line.index destination_station
  
  return_string = """You must travel through the following stops on the #{start_line_input.to_s} line: 
"""

  if starting_line == destination_line # We're only travelling on one line
    if start_index < destination_index
      return_string += starting_line[start_index..destination_index].join(', ')
    else
      return_string += starting_line[destination_index..start_index].reverse.join(', ')
    end
  else # Our trip travels through 2 lines
    if start_index < union_start  # Union is further down the line from where we start
      return_string += starting_line[start_index...union_start].join(', ')
      return_string += "\nChange at Union Square\nYour journey continues through to the following stops on the #{destination_line_input.to_s} line:"
      if destination_index < union_end # We travel back down the line to our destination
        return_string += "\n#{destination_line[destination_index...union_end].reverse.join(', ')}"
      else # We travel further up the line to our destination
        return_string += "\n#{destination_line[(union_end + 1)..destination_index].join(', ')}"
      end
    else # Union ISNT further down the line from where we start
      return_string += starting_line[(union_start + 1)..start_index].reverse.join(', ')
      return_string += "\nChange at Union Square\nYour journey continues through to the following stops on the #{destination_line_input.to_s} line:"
      if destination_index < union_end # We travel back down the line to our destination
        return_string += "\n#{destination_line[destination_index...union_end].reverse.join(', ')}"
      else # We travel further up the line to our destination
        return_string += "\n#{destination_line[(union_end + 1)..destination_index].join(', ')}"
      end
    end
  end
  puts return_string
end
plan_trip

File: Farid Masjedi/week07/mta/src/components/MTA.js
import React, {Component} from 'react'
import Search from './Search';
import Result from './Result';

class MTA extends Component {
  constructor(){
    super();

    this.state = {
      firstStation:'',
      firstLine:'',
      lastStation: '',
      secondLine: '',
      result:'',
      tripArrayInFirstLine:[],
      tripArrayInSecondLine:[],
      lines: {
        N: [ 'Times Square' , '34th' , '28th' , '23rd' , 'Union Square' , '8th' ],
        L: [ '8th' , '6th' , 'Union Square' , '3rd' , '1st' ],
        6: [ 'Grand Central' , '33rd' , '28th' , '23rd' , 'Union Square' , 'Astor Place' ]
      },
      unionStation: 'Union Square',
      flag: 'visible'
    }
  }

  generateTrip = (firstStation,firstLine,lastStation,secondLine) => {

    this.setState({firstStation:'',
                   firstLine:'',
                   lastStation: '',
                   secondLine: '',
                   result:'',
                   tripArrayInFirstLine:[],
                   tripArrayInSecondLine:[],
                   flag: 'visible'
                 })

   // Check: if the lines existed or not.
   // Exp: planTrip( 'West', '5th','N', 'Times Square')
   if (!Object.keys(this.state.lines).includes(firstLine)) {
     return this.setState({result:`There is no ${firstLine} line.`})
   } else if (!Object.keys(this.state.lines).includes(secondLine)){
     return this.setState({result:`There is no ${secondLine} line.`})
   }

   // // check: if the stops are in the lines or not.
   // // Exp: planTrip( 'N', '34th','L', 'Times Square')
   if (!this.state.lines[firstLine].includes(firstStation)) {
     return this.setState({result:`There is no ${firstStation} station in line ${firstLine}`})
   } else if (!this.state.lines[secondLine].includes(lastStation)) {
     return this.setState({result:`There is no ${lastStation} station in line ${secondLine}`});
   };

   // check: if one of the stops is 'Union Square', but two different -
   // - lines has been inputed.
   // Exp: planTrip( 'N', 'Union Square','L', 'Times Square')
   if (firstStation === this.state.unionStation) {
     firstLine = secondLine;
   }else if (lastStation === this.state.unionStation) {
     secondLine = firstLine;
   }

   // Structure of the function.
   let result = '';
   let startStop = this.state.lines[firstLine].indexOf(firstStation);
   let lastStop = this.state.lines[secondLine].indexOf(lastStation);
   let unionIndex1 = this.state.lines[firstLine].indexOf(this.state.unionStation);
   let unionIndex2 = this.state.lines[secondLine].indexOf(this.state.unionStation);

   let totalstops = 0;
   let travel = [];

   // Check if the trip is inside one of the lines.
   if (firstLine === secondLine) {
     let first = 0;
     let last = 0;

     totalstops = Math.abs(lastStop-startStop);
     first = Math.min(lastStop,startStop);
     last = Math.max(lastStop,startStop);

     travel = this.state.lines[firstLine].slice(first,last+1);
     if (startStop>lastStop){
       travel.reverse();
     }
     this.setState({tripArrayInFirstLine:travel})
     travel.shift();

     let result = `You must travel through the following stops on the line ${firstLine} : ${travel.join(', ')}.`;
     result = result + `\n\n${totalstops} stops in total.`;
     this.setState({result:result});

   }else{

     let firstS = 0;
     let firstE = 0;
     let secondS = 0;
     let secondE = 0;

     // if changing the line is needed.
     totalstops = Math.abs(startStop-unionIndex1)+Math.abs(unionIndex2-lastStop);
     firstS = Math.min(startStop,unionIndex1);
     firstE = Math.max(startStop,unionIndex1)
     secondS = Math.min(lastStop,unionIndex2);
     secondE = Math.max(lastStop,unionIndex2);

     travel = this.state.lines[firstLine].slice(firstS,firstE+1)

     if (startStop>unionIndex1){
       travel.reverse();
     }
     this.setState({tripArrayInFirstLine:[...travel]})
     travel.shift();
     let result = `You must travel through the following stops on the line ${firstLine} : ${travel.join(', ')}.`;
     result = result + `\n\nChange at ${this.state.unionStation} and go to line ${secondLine}.`;

     travel = [];
     travel = this.state.lines[secondLine].slice(secondS,secondE+1);

     if (unionIndex2>lastStop){
       travel.reverse();
     }

     travel.shift();
     this.setState({tripArrayInSecondLine:[...travel]})


     result = result + `\n\nYour journey continues through the following stops: ${travel.join(', ')}.`;
     result = result + `\n\n${totalstops} stops in total.`;
     this.setState({result:result});
   }

   this.setState({firstStation:firstStation,
                  firstLine:firstLine,
                  lastStation:lastStation,
                  secondLine:secondLine,
                  flag: 'hide'
                });
  }

  render(){

    return(

      <div className='container'>
      
        <Search onSubmit={ this.generateTrip }/>

        <div>

          <Result
            firstStation={this.state.firstStation}
            firstLine={this.state.firstLine}
            lastStation={this.state.lastStation}
            secondLine={this.state.secondLine}
            result ={this.state.result}
          />

          <div className='route'>
            {(this.state.tripArrayInFirstLine.length) ?
                <div className="trip">

                  <h1>Line {this.state.firstLine}</h1>
                  {this.state.tripArrayInFirstLine.map( (stop) => (
                        <div className='stops'>
                          <div className='circle-trip-line1'></div>
                          <p>{stop}</p>
                        </div>
                  ))}

                </div>

            : null}

            {(this.state.tripArrayInSecondLine.length) ?
                <div className="trip">

                  <h1>Line {this.state.secondLine}</h1>
                  {this.state.tripArrayInSecondLine.map( (stop) => (
                        <div className='stops'>
                          <div className='circle-trip-line2'></div>
                          <p>{stop}</p>
                        </div>
                  ))}

                </div>

            : null}
          </div>

          {Object.keys(this.state.lines).map( (line) => (

            <div className={this.state.flag}>
              <h1>Line {line}</h1>
              {this.state.lines[line].map( (stop) => (
                <div className='stops'>
                  <div className='circle'></div>
                  <p>{stop}</p>
                </div>
              ))}

            </div>

          ))}

        </div>

      </div>
    )
  }
}

export default MTA;
